components/brave_wallet/common/fil_address.cc

```diff
@@ -0,0 +1,180 @@
+         network == mojom::kFilecoinMainnet;
+}
+
+absl::optional<mojom::FilecoinAddressProtocol> ToProtocol(char input) {
+  if ((input - '0') == static_cast<int>(mojom::FilecoinAddressProtocol::BLS))
+    return mojom::FilecoinAddressProtocol::BLS;
+  if ((input - '0') ==
+      static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1))
+    return mojom::FilecoinAddressProtocol::SECP256K1;
+  return absl::nullopt;
+}
+
+}  // namespace
+
+FilAddress::FilAddress(const std::vector<uint8_t>& bytes,
+                       mojom::FilecoinAddressProtocol protocol,
+                       const std::string& network)
+    : protocol_(protocol), network_(network), bytes_(bytes) {
+  DCHECK(IsValidNetwork(network));
+}
+FilAddress::FilAddress() = default;
+FilAddress::FilAddress(const FilAddress& other) = default;
+FilAddress::~FilAddress() = default;
+
+bool FilAddress::IsEqual(const FilAddress& other) const {
+  return bytes_.size() == other.bytes_.size() &&
+         std::equal(bytes_.begin(), bytes_.end(), other.bytes_.begin()) &&
+         protocol_ == other.protocol_ && network_ == other.network_;
+}
+
+bool FilAddress::operator==(const FilAddress& other) const {
+  return IsEqual(other);
+}
+
+bool FilAddress::operator!=(const FilAddress& other) const {
+  return !IsEqual(other);
+}
+
+// Decodes Filecoin BLS/SECP256K addresses within rules
+// https://spec.filecoin.io/appendix/address/#section-appendix.address.string
+// |------------|----------|---------|----------|
+// |  network   | protocol | payload | checksum |
+// |------------|----------|---------|----------|
+// | 'f' or 't' |  1 byte  | n bytes | 4 bytes  |
+// static
+FilAddress FilAddress::FromAddress(const std::string& address) {
+  if (address.size() != kAddressSizeBLS &&
+      address.size() != kAddressSizeSecp256K)
+    return FilAddress();
+
+  auto protocol = ToProtocol(address[1]);
+  if (!protocol)
+    return FilAddress();
+
+  std::string network{address[0]};
+  if (!IsValidNetwork(network))
+    return FilAddress();
+
+  std::string payload{
+      base32::Base32Decode(base::ToUpperASCII(address.substr(2)))};
+  if (payload.empty())
+    return FilAddress();
+
+  std::string key{payload.substr(0, payload.size() - kChecksumSize)};
+  std::vector<uint8_t> public_key(key.begin(), key.end());
+  return FilAddress::FromPublicKey(public_key, protocol.value(), network);
+}
+
+// Creates FilAddress from SECP256K uncompressed public key
+// with specified protocol  and network
+// https://spec.filecoin.io/appendix/address/#section-appendix.address.string
+// static
+FilAddress FilAddress::FromUncompressedPublicKey(
+    const std::vector<uint8_t>& uncompressed_public_key,
+    mojom::FilecoinAddressProtocol protocol,
+    const std::string& network) {
+  if (protocol == mojom::FilecoinAddressProtocol::BLS)
+    return FilAddress();
+  if (uncompressed_public_key.empty())
+    return FilAddress();
+  auto public_key = BlakeHash(uncompressed_public_key, kPublicKeySizeSecp256K);
```

Comment:
This is not the public key but blake2b-160(uncompressed_secp256k1_public_key) if I understand the document correctly.
And it's more confusing that kPublicKeySizeSecp256K you defined isn't the size of secp256k1 public key either but the size of hashed result.

Reply:
https://github.com/brave/brave-core/blob/master/components/brave_wallet/browser/internal/hd_key.cc#L404

Reply:
??

Reply:
done